autotest/alg/polygonize.py

```diff
@@ -181,15 +181,106 @@ def test_polygonize_4():
 
     # Confirm we get the set of expected features in the output layer.
 
-    expected_feature_number = 16
+    expected_feature_number = 17
     assert mem_layer.GetFeatureCount() == expected_feature_number
 
-    expect = [107, 123, 132, 115, 132, 115, 140, 148,
+    expect = [107, 123, 132, 115, 132, 115, 140, 132, 148,
               123, 140, 132, 156, 100, 101, 102, 103]
 
     tr = ogrtest.check_features_against_list(mem_layer, 'DN', expect)
 
     assert tr
 
+###############################################################################
+# Test a simple case with holes.
+
+
+def test_polygonize_5():
+
+    src_ds = gdal.Open('data/polygonize_in_3.grd')
+    src_band = src_ds.GetRasterBand(1)
+
+    # Create a memory OGR datasource to put results in.
+    mem_drv = ogr.GetDriverByName('Memory')
+    mem_ds = mem_drv.CreateDataSource('out')
+
+    mem_layer = mem_ds.CreateLayer('poly', None, ogr.wkbPolygon)
+
+    fd = ogr.FieldDefn('DN', ogr.OFTInteger)
+    mem_layer.CreateField(fd)
+
+    # run the algorithm.
+    result = gdal.Polygonize(src_band, None, mem_layer, 0)
+    assert result == 0, 'Polygonize failed'
+
+    # Confirm we get the set of expected features in the output layer.
+
+    expected_feature_number = 4
+    assert mem_layer.GetFeatureCount() == expected_feature_number
+
+    expect = [1, 0, 1, 0]
+    expect_wkt = [
+        'POLYGON ((0 4,0 0,4 0,4 4,0 4),(1 3,1 1,3 1,3 2,2 2,2 3,1 3))',
+        'POLYGON ((1 3,1 2,2 2,2 3,1 3))',
+        'POLYGON ((1 2,1 1,2 1,2 2,1 2))',
```

Comment:
This 3rd polygon shouldn't be emitted. It is within the first one

Reply:
It is not within the first polygon. That has a L-shaped hole that is filled with the three one pixel polygons. It is just that the third polygon could be merged with the first polygon because they both have pixel value=1. Now the result is indeed topologically correct but as merged it would be even better.

![kuva](https://user-images.githubusercontent.com/1751612/103467592-49c4e280-4d59-11eb-91e2-dc6b2ab75a61.png)

Reply:
indeed, it's better to merge the 3rd polygon into first polygon. However. In this solution, it's a compromise between topo correctness and feature count correctness.